testsuite/tests/warnings/w44.ml

```diff
@@ -0,0 +1,26 @@
+(* TEST
+ flags = "-w +A";
+ expect;
+*)
+
+
+module M = struct let f () = 1 end;;
+[%%expect {|
+module M : sig val f : unit -> int end
+|}]
```

Comment:
Should give an example of the recursive module and class cases? For instance,
```ocaml
module K = struct
  module M = struct
    module type R =sig end
   end
end
module rec M: sig module type R open K module X:M.R end = M
```
and
```ocaml
module M = struct let x = 0 end
class c = object
  val x=0
  val y = M.( x ) + 1
end
```

Reply:
Thanks! I added the class example to the testsuite, but the module example does not work.

`Env.enter_unbound_module` is only used by mod_unbound` is only used in `Typemod.approx_modtype` which is a preliminary part of module checking and in particular passes `use:false` to its environment lookup functions. I understand vaguely that another type-checking pass later performs type-checking again with a less-approximated environment, and shadowing warning may then trigger for names that were unbound during the approximation.

I am not sure whether "this example of @Octachron does raise a shadowing warning" should be considered a bug or not (it is at the boundary where my reasoning above suggests that maybe it shouldn't raise), but this is a subtle issue that I think could deserve to be handled separately from #12494, so I propose to move on to other things instead of trying to add more code to "fix" this issue.

The remaining question is whether I should leave the example as a known-possibly-bad-but-we-haven't-really-decided case in the testsuite. I decided to keep it out of it.

If you feel very enthusiastic about this weird related issue, feel free to open a separate issue with the example. (Maybe a fix would be to mark *some* identifier queries in `approx_modtype` as uses?) I feel that we both have other things on our plate in the coming, uh, decades.

Reply:
I missed the fact that the shadowing warning will be raised later in the later cycles of the typing of recursive modules, sorry.

I completely agree that the recursive module example should be at best eventually be discussed later, maybe.